prototype/src/gpu/dust.hpp

```diff
@@ -239,10 +228,44 @@ class Dust {
   }
 
   void reset(const init_t data, const size_t step) {
-    size_t n_particles = _particles.size();
-    _particles.clear();
+    const size_t n_particles = _particles.size();
+    initialise(data, step, n_particles);
+  }
+
+  // It's the callee's responsibility to ensure that index is in
+  // range [0, n-1]
+  void set_index(const std::vector<size_t>& index) {
+    _index = index;
+  }
+
+  // It's the callee's responsibility to ensure this is the correct length
+  void set_state(const std::vector<real_t>& state, bool is_matrix) {
+    const size_t n_particles = _particles.size();
+    const size_t n_state = n_state_full();
+    auto it = state.begin();
     for (size_t i = 0; i < n_particles; ++i) {
-      _particles.push_back(Particle<T>(data, step));
+      _particles[i].set_state(it);
+      if (is_matrix) {
+        it += n_state;
+      }
+    }
+  }
+
+  void set_step(const size_t step) {
+    const size_t n_particles = _particles.size();
+    for (size_t i = 0; i < n_particles; ++i) {
+      _particles[i].set_step(step);
+    }
+  }
+
+  void set_step(const std::vector<size_t>& step) {
```

Comment:
Why not do this with `as_matrix` as well?

Reply:
Or perhaps `as_vector` here

Reply:
Sorry, I am unclear what is being asked here

Reply:
(This was about a single set_step that switched; to mirror the dust interface and avoid creating length-1 vectors opting to not do this)